jax/experimental/host_callback.py

```diff
@@ -165,27 +167,38 @@ def power3(x):
 XlaLocalClient = Any  # xla_extension.LocalClient
 
 
-def id_tap(tap_func: Callable, arg, *, result=None, **kwargs):
+T = TypeVar('T')
+U = TypeVar('U')
+_Transforms = Sequence[Tuple[str, Dict[str, Any]]]
+_TapFunc = Callable[[T, _Transforms], Any]
+
+@typing.overload
+def id_tap(tap_func: _TapFunc, arg: T) -> T:
+  ...
+
+@typing.overload
+def id_tap(tap_func: _TapFunc, arg: T, *, result: U) -> U:
+  ...
+
+def id_tap(tap_func, arg, *, result=None):
   """Host-callback tap primitive, like identity function with a call to ``tap_func``.
 
   **Experimental: please give feedback, and expect changes!**
 
   ``id_tap`` behaves semantically like the identity function but has the
-  side-effect
-  that a user-defined Python function is called with the runtime values of the
-  argument.
+  side-effect that a user-defined Python function is called with the runtime
+  values of the argument.
 
   Args:
-    * tap_func: the tap function to call.
+    * tap_func: tap function to call like ``tap_func(arg, transforms)``, with
+      ``arg`` as described below and where ``transforms`` is sequence of applied
+      JAX transformations in the form ``(name, params)``.
     * arg: the argument passed to the tap function, can be a pytree of JAX
       types.
     * result: if given, specifies the return value of ``id_tap``. This value is
       not passed to the tap function, and in fact is not sent from the device to
       the host. If the ``result`` parameter is not specified then the return
       value of ``id_tap`` is ``arg``.
-    * kwargs: will be passed directly to the tap function. Can be anything that
```

Comment:
I am not sure it is a good idea to remove this feature. What do you have in mind as a replacement for the ability to pass auxiliary data to the tap function (data that should not be part of the compiled code, e.g., strings)?

Reply:
See my comment in the main issue.

Reply:
Sounds good.